frame/democracy/src/lib.rs

```diff
@@ -0,0 +1,2084 @@
+		/// \[proposal_hash, provider, deposit, reaper\]
+		PreimageReaped(T::Hash, T::AccountId, BalanceOf<T>, T::AccountId),
+		/// A proposal \[hash\] has been blacklisted permanently.
+		Blacklisted(T::Hash),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Value too low
+		ValueLow,
+		/// Proposal does not exist
+		ProposalMissing,
+		/// Cannot cancel the same proposal twice
+		AlreadyCanceled,
+		/// Proposal already made
+		DuplicateProposal,
+		/// Proposal still blacklisted
+		ProposalBlacklisted,
+		/// Next external proposal not simple majority
+		NotSimpleMajority,
+		/// Invalid hash
+		InvalidHash,
+		/// No external proposal
+		NoProposal,
+		/// Identity may not veto a proposal twice
+		AlreadyVetoed,
+		/// Preimage already noted
+		DuplicatePreimage,
+		/// Not imminent
+		NotImminent,
+		/// Too early
+		TooEarly,
+		/// Imminent
+		Imminent,
+		/// Preimage not found
+		PreimageMissing,
+		/// Vote given for invalid referendum
+		ReferendumInvalid,
+		/// Invalid preimage
+		PreimageInvalid,
+		/// No proposals waiting
+		NoneWaiting,
+		/// The given account did not vote on the referendum.
+		NotVoter,
+		/// The actor has no permission to conduct the action.
+		NoPermission,
+		/// The account is already delegating.
+		AlreadyDelegating,
+		/// Too high a balance was provided that the account cannot afford.
+		InsufficientFunds,
+		/// The account is not currently delegating.
+		NotDelegating,
+		/// The account currently has votes attached to it and the operation cannot succeed until
+		/// these are removed, either through `unvote` or `reap_vote`.
+		VotesExist,
+		/// The instant referendum origin is currently disallowed.
+		InstantNotAllowed,
+		/// Delegation to oneself makes no sense.
+		Nonsense,
+		/// Invalid upper bound.
+		WrongUpperBound,
+		/// Maximum number of votes reached.
+		MaxVotesReached,
+		/// Maximum number of proposals reached.
+		TooManyProposals,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		/// Weight: see `begin_block`
+		fn on_initialize(n: T::BlockNumber) -> Weight {
+			Self::begin_block(n)
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Propose a sensitive action to be taken.
+		///
+		/// The dispatch origin of this call must be _Signed_ and the sender must
+		/// have funds to cover the deposit.
+		///
+		/// - `proposal_hash`: The hash of the proposal preimage.
+		/// - `value`: The amount of deposit (must be at least `MinimumDeposit`).
+		///
+		/// Emits `Proposed`.
+		///
+		/// Weight: `O(p)`
+		#[pallet::weight(T::WeightInfo::propose())]
+		pub fn propose(
+			origin: OriginFor<T>,
+			proposal_hash: T::Hash,
+			asset_id: T::AssetId,
+			#[pallet::compact] value: BalanceOf<T>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			ensure!(value >= T::MinimumDeposit::get(), Error::<T>::ValueLow);
+			let id = ProposalId { hash: proposal_hash, asset_id };
```

Comment:
weird for now, asset_id is not used anywere except hash. it is not validate or else.

Reply:
Yes, this is used only to generate proposal id, along with hash.